Update 2022-12-12-async-manual.md (#2)
krzysztofslusarski/krzysztofslusarski.github.io: _posts/2022-12-12-async-manual.md

    return result;
}
```
Both methods ```matrixMultiplySlow``` and ```matrixMultiplyFaster``` have the same complexity O(n^3). So
why is one faster than the other? Well, if you want to understand exactly how a CPU-intensive algorithm works, 
you need to understand how the CPU works, which is far away from the topic of this post. Be aware that if
you want to optimize such algorithms; you will probably need at least one of the following:

- Knowledge of CPU architecture
- Top-Down performance analysis methodology
- Looking at the ASM of generated methods  

> Many Java programmers forget that all the execution is done on the CPU. Java needs to use tak ASM to run on the CPU. That's

Reviewer: ```suggestion
Many Java programmers forget that all the execution is done on the CPU. Java needs to use ASM to run on the CPU. That's
```

born. Let's assume for simplicity that eden is a continuous slice of memory. The very efficient
way of allocation in such a case is called **bumping pointer**. We keep a pointer to the first 
address of the free space:

![alt text](/assets/async-demos/alloc-1.png "alloc")

When we do ```new Object()```, we simply calculate its size, locate 
the next free address and bump the pointer by ```sizeof(Object)```:

![alt text](/assets/async-demos/alloc-2.png "alloc")

But there is one major problem with that technique: We have to synchronize the object allocation if we have more than one thread that can 
create new objects in parallel, but this is quite costly. We solve this by giving each thread a portion of eden dedicated to only that thread. This portion 
is called **TLAB** - thread-local allocation buffer. With this, each thread can use **bumping pointer** at its TLAB safely and in parallel.

Introducing TLABs creates two more issues that the JVM needs to deal with:

- a thread can allocate an object, but there is not enough space in its TLAB - the JVM creates a new TLAB if there is still space in eden

> - a thread can allocate a big object, so it's not optimal to use the TLAB mechanism - the JVM will use the _slow path_ of the allocation that allocates the object directly in eden

Reviewer: ```suggestion
- a thread can allocate a big object, so it's not optimal to use the TLAB mechanism - the JVM will use the _slow path_ of the allocation that allocates the object directly in eden or in the old generation
```

> - a thread can allocate a big object, so it's not optimal to use the TLAB mechanism - the JVM will use the _slow path_ of the allocation that allocates the object directly in eden

Reviewer: G1 Humongous objects are created directly in the old-gen, forgot about it before.

You can find **three** sizes in such an entry **A->B(C)** that are:
* **A** - used size of a heap before _GC cycle_
* **B** - used size of a heap after _GC cycle_
* **C** - the current size of a whole heap

If we take the **B** value from each collection and put it on a chart, we can generate the 
_Heap after GC_ chart. We can use such a chart to then detect if we have a memory leak:
If a chart looks like those (from a **7 days** period):

![alt text](/assets/monday-2/1.jpg "1")

![alt text](/assets/monday-2/4.jpg "4")

then there is **no memory leak**. The _garbage collector_ can clean up the heap to 
the same level every day. The chart with a memory leak looks like the following one:

![alt text](/assets/monday-2/2.jpg "2")

> These spikes to the roof are _too-space exhausted_ situations in the **G1** algorithm; those are not _OutOfMemoryErrors_. After each of those spikes, there was a

Reviewer: ```suggestion
These spikes to the roof are _to-space exhausted_ situations in the **G1** algorithm; those are not _OutOfMemoryErrors_. After each of those spikes, there was a
```

}
```

Just set ```writableStackTrace``` to ```false```. It will be rather ugly but faster.

### G1GC humongous allocation
{: #methods-g1ha }

We already saw how to detect humongous objects with allocation mode. Since
async-profiler can also instrument JVM code, and allocations of a humongous objects are 
nothing else than invocations of C++ code, we can take advantage of that.

> If you want to check where humongous objects are allocated, you can use method mode

Reviewer: ```suggestion
If you want to check where humongous objects are allocated, you can use native functions mode
```

Reviewer: Method mode is not the same as native functions. It's a bit undocumented, but Andrei ask me to separate those two.

Author: Good to know

![alt text](/assets/async-demos/threads-2.png "threads")

Then you might want to check who is creating those short-living threads:

> We use async-profiler with the ```JVM_StartThread``` event in method mode for this purpose:

Reviewer: ```suggestion
We use async-profiler with the ```JVM_StartThread``` event in native functions mode for this purpose:
```
